File: java/74_SearchA2DMatrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if (matrix.length == 0 || matrix[0].length == 0) return false;
        
        int rows = matrix.length, cols = matrix[0].length;
        int left = 0, right = rows * cols - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            int i = mid / cols, j = mid % cols;
            
            if (matrix[i][j] == target) return true;
            else if (matrix[i][j] < target) left = mid + 1;
            else right = mid - 1;
        }
        return false;
    }
}

File: swift/74_SearchA2DMatrix.swift
class Solution {
    func searchMatrix(_ matrix: [[Int]], _ target: Int) -> Bool {
        if matrix.isEmpty || matrix[0].isEmpty { return false }
        
        var rows = matrix.count, cols = matrix[0].count
        var left = 0, right = rows * cols - 1
        while left <= right {
            var mid = (left + right) >> 1
            var i = mid / cols, j = mid % cols
            
            if matrix[i][j] == target { return true }
            else if matrix[i][j] > target { right = mid - 1 }
            else { left = mid + 1 }
        }
        return false
    }
}
